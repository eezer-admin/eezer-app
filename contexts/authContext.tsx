import { createContext, useEffect, useState } from 'react';

import { User } from '@src/domain/entities/User';
import { ClearUserUseCase } from '@usecases/user/ClearUserUseCase';
import { GetUserUseCase } from '@usecases/user/GetUserUseCase';
import { LoginUseCase } from '@usecases/user/LoginUseCase';
import { removeFromStorage as removeTransportLog } from '../services/TransportLogService';
import { remove as removeCurrentTransport } from '../services/TransportService';
import { AuthContextData } from '../types/Auth';

export const AuthContext = createContext<AuthContextData>({} as AuthContextData);

export const AuthProvider = (props) => {
  const [user, setUser] = useState<AuthContextData>();
  const [authLoaded, setAuthLoaded] = useState(false);

  // Determine if the user is already logged in when the app loads.
  // If there's a user in the local storage, assign it to the context
  // to make the app recognize it.
  useEffect(() => {
    new GetUserUseCase().execute().then((user: User | null) => {
      if (user) {
        setUser(user);
      }
      setAuthLoaded(true);
    });
  }, []);

  // Makes an actual login request towards the backend API.
  const login = async (username: string, password: string): Promise<User | null> => {
    return new LoginUseCase().execute(username, password).then((user: User) => {
      if (user) {
        setUser(user);

        return user;
      }

      return null;
    });
  };

  // Removes the user from the local storage and context.
  const logout = async () => {
    await removeTransportLog();
    await removeCurrentTransport();
    await new ClearUserUseCase().execute();

    setUser(null);
  };

  const isLoggedIn = (): boolean => {
    return !!user?.access_token;
  };

  return (
    <AuthContext.Provider value={{ user, authLoaded, login, logout, isLoggedIn }}>
      {props.children}
    </AuthContext.Provider>
  );
};
