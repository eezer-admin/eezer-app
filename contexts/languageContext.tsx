import { createContext, useEffect, useState } from 'react';

import i18n, { defaultLanguage } from '../localization/Localization';

export const LanguageContext = createContext<string>(defaultLanguage());

export const LanguageProvider = (props) => {
  const [language, setLanguage] = useState(props.language);

  // Update the i18n locale config before re-rendering the app.
  useEffect(() => {
    i18n.locale = language;
  }, [language]);

  return (
    <LanguageContext.Provider value={[language, setLanguage]}>
      {props.children}
    </LanguageContext.Provider>
  );
};
