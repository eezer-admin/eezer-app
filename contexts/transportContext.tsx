import { createContext, useEffect, useState } from 'react';

import { Transport } from '@src/domain/entities/Transport';
import { GetTransportUseCase } from '@usecases/transport/GetTransportUseCase';
import TransportModel from '../models/TransportModel';
import { get as getFromStorage, persist, remove } from '../services/TransportService';
import { TransportContextData } from '../types/Transports';

export const TransportContext = createContext<TransportContextData>({} as TransportContextData);

export const TransportProvider = (props) => {
  const [data, setData] = useState<TransportContextData>();
  const [transport, setTransport] = useState<Transport | null>(null);

  // Check if the user already has an ongoing transport in the local storage
  // when opening the app. If so, navigate directly to it.
  useEffect(() => {
    new GetTransportUseCase().execute().then((transport: Transport | null) => {
      if (transport) {
        setData(transport);
        setTransport(transport);
      }
    });
  }, []);

  const get = async (): Promise<TransportModel | null> => {
    return getFromStorage();
  };

  const start = async (reason: string, vehicle_id: number): Promise<TransportModel> => {
    const transport = new TransportModel(null);
    transport.start(reason);
    transport.setVehicleId(vehicle_id);
    setData(transport);

    return persist(transport);
  };

  const completeTransport = async (): Promise<void> => {
    setData(null);

    return remove();
  };

  const save = async (transport: TransportModel): Promise<TransportModel> => {
    return persist(transport).then(() => {
      setData(transport);

      return transport;
    });
  };

  return (
    <TransportContext.Provider
      value={{
        transport,
        setTransport,
        data,
        get,
        save,
        start,
        completeTransport,
      }}>
      {props.children}
    </TransportContext.Provider>
  );
};
