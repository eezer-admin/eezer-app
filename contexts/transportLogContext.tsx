import { createContext, useEffect, useState } from 'react';

import { add, get as getFromStorage } from '../services/TransportLogService';
import { TransportLog, TransportLogContextData } from '../types/Transports';

export const TransportLogContext = createContext<TransportLogContextData>({} as TransportLogContextData);

export const TransportLogProvider = (props) => {
  const [data, setData] = useState<TransportLogContextData>();

  useEffect(() => {
    getFromStorage().then((log: TransportLog) => {
      if (log) {
        setData(log);
      }
    });
  }, []);

  return <TransportLogContext.Provider value={[data, setData]}>{props.children}</TransportLogContext.Provider>;
};
