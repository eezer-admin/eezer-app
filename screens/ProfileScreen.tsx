import * as React from 'react';
import { useContext } from 'react';
import { View, Text } from 'react-native';

import { AuthContext } from '../contexts/authContext';
import Styles from '../styles/Styles';

export default function ProfileScreen() {
  const auth = useContext(AuthContext);

  return (
    <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
      <Text
        style={{
          marginBottom: Styles.margins.medium,
          ...Styles.text.default,
        }}>
        {auth.user.first_name} {auth.user.last_name}
      </Text>
      <Text style={{ ...Styles.text.default }}>{auth.user.email}</Text>
    </View>
  );
}
