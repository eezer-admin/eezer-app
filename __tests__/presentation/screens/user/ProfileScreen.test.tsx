import { describe, expect, it } from '@jest/globals';
import ProfileScreen from '@presentation/screens/user/ProfileScreen';
import { SafeAreaProvider } from 'react-native-safe-area-context';
import renderer from 'react-test-renderer';

import { AuthProvider } from '../../../../contexts/authContext';

describe('ProfileScreen', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <SafeAreaProvider>
          <AuthProvider>
            <ProfileScreen />
          </AuthProvider>
        </SafeAreaProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
