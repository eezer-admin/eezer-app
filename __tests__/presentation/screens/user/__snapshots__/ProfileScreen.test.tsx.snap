// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileScreen renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
/>
`;
